src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/IncludeFlagUtils.cs

```diff
@@ -53,15 +53,39 @@ internal static Dictionary<string, LibraryIncludeFlags> FlattenDependencyTypes(
             {
                 if (result.ContainsKey(dependency.Name))
                 {
-                    result[dependency.Name] = dependency.IncludeType;
+                    if (IsDependencyPruned(dependency, specFramework?.PackagesToPrune))
+                    {
+                        result[dependency.Name] = LibraryIncludeFlags.None;
+                    }
+                    else
+                    {
+                        result[dependency.Name] = dependency.IncludeType;
+                    }
                 }
                 else
                 {
-                    result.Add(dependency.Name, dependency.IncludeType);
+                    if (IsDependencyPruned(dependency, specFramework?.PackagesToPrune))
+                    {
+                        result[dependency.Name] = LibraryIncludeFlags.None;
+                    }
+                    else
+                    {
+                        result.Add(dependency.Name, dependency.IncludeType);
+                    }
                 }
             }
 
             return result;
+
+            static bool IsDependencyPruned(LibraryDependency dependency, IReadOnlyDictionary<string, PrunePackageReference> packagesToPrune)
+            {
+                if (packagesToPrune?.TryGetValue(dependency.Name, out PrunePackageReference packageToPrune) == true
+                    && dependency.LibraryRange.VersionRange.Satisfies(packageToPrune.VersionRange.MaxVersion))
+                {
+                    return true;
+                }
+                return false;
+            }
```

Comment:
We can reduce verbosity and make it more concise:

```suggestion
            static bool IsDependencyPruned(LibraryDependency dependency, IReadOnlyDictionary<string, PrunePackageReference> packagesToPrune)
            {
                return packagesToPrune?.TryGetValue(dependency.Name, out var packageToPrune) == true
                    && dependency.LibraryRange.VersionRange.Satisfies(packageToPrune.VersionRange.MaxVersion);
            }
```

Reply:
There was similar question in an previous PR, that implementation slows down debugging, since now you need to set a conditional breakpoint vs just setting a breakpoint. 
Functionally they're the same, so no perf benefit. 

From a readability perspective, I've always seen them as equivalent as well, especially in this case. 

If you feel strongly I can change it, especially both you and Nigusu brought it up (although I'm guessing @zivkan will lean towards the current version, as that's something we've discussed previously)

Reply:
yes, I personally dislike return statements that do anything other than return a const or a variable, because I often want to inspect the value when debugging.  The compiled code is the same, so I just see it as unhelpful code golf 😆 

I'm fine with `bool result = ...` followed by `return result;` if we want to remove the if statement. But I just don't like combining the two into a single statement.